### PR TITLE
Add disable domain port auto discovery by assignment client option.

### PR DESCRIPTION
**Pull request by digisomni**
This should (theoretically) fix https://github.com/vircadia/vircadia/issues/1250.
It does not update the pkg-script for creating .debs however, a follow-up PR for that will be needed.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Interface-PR1251-1c72e46-_2d4da7ba.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1c72e46-_2d4da7ba.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1c72e46-_2d4da7ba.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1c72e46-.tar_2d4da7ba.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1c72e46-.tar_2d4da7ba.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1c72e46-_2d4da7ba.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1c72e46_2d4da7ba.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Quest-Alpha-PR1251-1c72e46_2d4da7ba.apk

**Comment by digisomni** [comment]
"deadhand" tested this on an Ubuntu 18.04 LTS VM and confirmed successfully that the problem is no longer present.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Interface-PR1251-3010da0-_42a4e3b1.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-3010da0-_42a4e3b1.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-3010da0-.tar_42a4e3b1.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-3010da0-.tar_42a4e3b1.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-3010da0-_42a4e3b1.sh

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Quest-Alpha-PR1251-3010da0_42a4e3b1.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-3010da0_42a4e3b1.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-3010da0-_42a4e3b1.exe

File: libraries/networking/src/NodeList.cpp [hunk 1]
```diff
@@ -377,7 +377,7 @@ void NodeList::sendDomainServerCheckIn() {
             // if so we need to make sure we have an up-to-date local port in case it restarted
 
             if (domainSockAddr.getAddress() == QHostAddress::LocalHost
-                || hostname == "localhost") {
+                || hostname == "localhost" && _domainPortAutoDiscovery == true) {
```
Comment (odysseus654) [inline libraries/networking/src/NodeList.cpp:380]: Mixing && and || here is confusing and wrong.  Please add parenthesis here, specifically as (a || b) && c

File: pkg-scripts/vircadia-assignment-client.service [hunk 2]
```diff
@@ -12,7 +12,7 @@ User=vircadia
 Group=vircadia
 #LimitCORE=infinity
 #ExecStart=/opt/vircadia/assignment-client -n 6
-ExecStart=/opt/vircadia/assignment-client --min 6 --max 20
+ExecStart=/opt/vircadia/assignment-client --min 6 --max 20 --disable-domain-port-auto-discovery
```
Comment (odysseus654) [inline pkg-scripts/vircadia-assignment-client.service:15]: Please remove this change; the service without the @ is supposed to operate without the multi-domain workarounds.

**Review by odysseus654 — CHANGES_REQUESTED** [review]

File: assignment-client/src/AssignmentClient.cpp [hunk 3]
```diff
@@ -89,6 +90,13 @@ AssignmentClient::AssignmentClient(Assignment::Type requestAssignmentType, QStri
     _assignmentServerSocket.setObjectName("AssignmentServer");
     nodeList->setAssignmentServerSocket(_assignmentServerSocket);
 
+    if (disableDomainPortAutoDiscovery == true) {
+        _disableDomainPortAutoDiscovery = disableDomainPortAutoDiscovery;
+        qCDebug(assignment_client) << "Disabling domain port auto discovery by the assignment client due to parsed command line parameter.";
+    }
+
+    nodeList->setDomainPortAutoDiscovery(_disableDomainPortAutoDiscovery);
```
Comment (odysseus654) [inline assignment-client/src/AssignmentClient.cpp:98]: This value seems to be backwards, shouldn't it be set to !_disableDomainPortAutoDiscovery?

**Review by odysseus654 — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1f56203-_88b9b746.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1f56203-.tar_88b9b746.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1f56203-_88b9b746.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1f56203-.tar_88b9b746.gz

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Interface-PR1251-1f56203-_88b9b746.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1f56203_88b9b746.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Quest-Alpha-PR1251-1f56203_88b9b746.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-1f56203-_88b9b746.exe

File: assignment-client/src/AssignmentClientMonitor.cpp [hunk 4]
```diff
@@ -198,6 +201,9 @@ void AssignmentClientMonitor::spawnChildClient() {
         _childArguments.append("--" + ASSIGNMENT_TYPE_OVERRIDE_OPTION);
         _childArguments.append(QString::number(_requestAssignmentType));
     }
+    if (_disableDomainPortAutoDiscovery != false) {
```
Comment (ctrlaltdavid) [inline assignment-client/src/AssignmentClientMonitor.cpp:204]:
```suggestion
    if (!_disableDomainPortAutoDiscovery) {
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a661a25-.tar_35948d74.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a661a25-.tar_35948d74.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a661a25-_35948d74.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Interface-PR1251-a661a25-_35948d74.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a661a25-_35948d74.exe

**Comment by daleglass** [comment]
One doubt I have here is whether this is right way to go, perhaps we need to look into what needs to be done to properly support multiple domains on the same server?

**Comment by daleglass** [comment]
Also, perhaps this should be in the domain settings page?

**Comment by ctrlaltdavid** [comment]
How does this relate to the automatic network setting in the server settings?
image

**Comment by digisomni** [comment]
Uhh... All good questions. So, I'm not entirely sure. It's a domain server seeking issue, so maybe there should be a networking setting for this instead of just putting it as a parameter. The parameter should remain for the case of a service needing to configure this.

I think I will revise it then if possible. It's been a while since I looked at this.

**Comment by odysseus654** [comment]
The auto-discovery settings being discussed here are mostly applicable for a user installing it on their home machine as a sandbox or as a locally-served domain to have it work with minimal configuration; it's there to get it up and running and also to auto-heal a broken configuration in case the domain server restarts with a different port number.  I would say that it is *not* related to any metaverse managed stuff.  It is also not as useful if the assignment clients are given explicit connection parameters by whatever is launching them (assuming those parameters don't change in the middle of execution).

The autodiscovery would likely be useless/spamming as well if the assignment client were trying to connect to a domain on a different machine (although I think this check is skipped if that's detected to be the case), and if there's multiple domain servers on a single machine and the assignment clients are somehow able to see a domain server other than the one they're connected to they may unexpectedly "jump ship" between servers.

File: assignment-client/src/AssignmentClient.cpp [hunk 5]
```diff
@@ -90,6 +91,13 @@ AssignmentClient::AssignmentClient(Assignment::Type requestAssignmentType, QStri
     _assignmentServerSocket.setObjectName("AssignmentServer");
     nodeList->setAssignmentServerSocket(_assignmentServerSocket);
 
+    if (disableDomainPortAutoDiscovery == true) {
```
Comment (odysseus654) [inline assignment-client/src/AssignmentClient.cpp:94]:
```suggestion
    if (disableDomainPortAutoDiscovery) {
```

File: libraries/networking/src/NodeList.cpp [hunk 6]
```diff
@@ -376,8 +376,8 @@ void NodeList::sendDomainServerCheckIn() {
             // is this our localhost domain-server?
             // if so we need to make sure we have an up-to-date local port in case it restarted
 
-            if (domainSockAddr.getAddress() == QHostAddress::LocalHost
-                || hostname == "localhost") {
+            if ((domainSockAddr.getAddress() == QHostAddress::LocalHost || hostname == "localhost")
+                && _domainPortAutoDiscovery == false) {
```
Comment (odysseus654) [inline libraries/networking/src/NodeList.cpp:380]:
```suggestion
                && _domainPortAutoDiscovery) {
```

File: libraries/networking/src/NodeList.h [hunk 7]
```diff
@@ -71,6 +71,8 @@ class NodeList : public LimitedNodeList {
     void setAssignmentServerSocket(const SockAddr& serverSocket) { _assignmentServerSocket = serverSocket; }
     void sendAssignment(Assignment& assignment);
 
+    void disableDomainPortAutoDiscovery(bool disabled = false) { _domainPortAutoDiscovery = disabled; };
```
Comment (odysseus654) [inline libraries/networking/src/NodeList.h:74]:
```suggestion
    void disableDomainPortAutoDiscovery(bool disabled = false) { _domainPortAutoDiscovery = !disabled; };
```

File: libraries/networking/src/NodeList.h [hunk 8]
```diff
@@ -181,6 +183,7 @@ private slots:
     bool _requestsDomainListData { false };
 
     bool _sendDomainServerCheckInEnabled { true };
+    bool _domainPortAutoDiscovery { false };
```
Comment (odysseus654) [inline libraries/networking/src/NodeList.h:186]:
```suggestion
    bool _domainPortAutoDiscovery { true };
```

File: pkg-scripts/vircadia-assignment-client.service [hunk 9]
```diff
@@ -1,5 +1,5 @@
 [Unit]
-Description=Assignment client service for Vircadia server
+Description=Assignment client service for the Vircadia Server
```
Comment (odysseus654) [inline pkg-scripts/vircadia-assignment-client.service:2]: I can't add it as a suggestion in GitHub here, but why didn't you make the change in the other file as well?

**Review by odysseus654 — CHANGES_REQUESTED** [review]
Okay another code review, some change requests:
- (high priority) the variable _domainPortAutoDiscovery on NodeList as written is confusing, either change the name or change the value.  The suggested changes are for the value
- Dropped a useless reference to "== true"
- The opposite of getLocalServerPortFromSharedMemory is putLocalPortIntoSharedMemory, is there any value in having the domain controller not broadcasting the local port number as well as the assignment clients not listening for it?

File: pkg-scripts/vircadia-assignment-client.service (hunk 9 above)
Comment (digisomni) [reply pkg-scripts/vircadia-assignment-client.service:2]: Probably because I missed it! I'll fix it.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-66aca25-.tar_e4cb87ba.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-66aca25-.tar_e4cb87ba.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-66aca25-_e4cb87ba.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Interface-PR1251-66aca25-_e4cb87ba.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-66aca25-_e4cb87ba.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-66aca25-_e4cb87ba.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a2faa9a-_76010bf6.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a2faa9a-.tar_76010bf6.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a2faa9a-.tar_76010bf6.gz

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-Interface-PR1251-a2faa9a-_76010bf6.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a2faa9a-_76010bf6.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1251/Vircadia-PR1251-a2faa9a-_76010bf6.exe

**Comment by odysseus654** [comment]
I have no further obvious feedback based on code inspection alone, think I've exhausted everything I could think of.

**Comment by digisomni** [comment]
Works fine with and without the parameter on Windows 10 as well it seems.

**Pull request merged**